File: ios/Yizhidao/Features/Reading/AIAnswerFormatter.swift
import Foundation

/// AI 常把解读写成一整段，手机上读起来很挤。这里只在展示层分段，不动存下来的原文：
/// 先按换行切块，再只在句末（。！？…）断开，累计约 60 字成段，末段过短并回上一段。
enum AIAnswerFormatter {
    static func paragraphs(in raw: String) -> [String] {
        let blocks = raw
            .replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        return blocks.flatMap(paragraphs(inBlock:))
    }

    private static let softLimit = 60
    private static let minTailLength = 24
    private static let enders: Set<Character> = ["。", "！", "？", "…", "!", "?"]
    private static let closers: Set<Character> = ["」", "』", "”", "’", "）", ")", "》", "〉", "】", "、"]

    private static func paragraphs(inBlock block: String) -> [String] {
        guard block.count > softLimit else { return [block] }
        var result: [String] = []
        var current = ""
        for sentence in sentences(in: block) {
            current += sentence
            if current.count >= softLimit {
                result.append(current)
                current = ""
            }
        }
        if !current.isEmpty {
            if current.count < minTailLength, let last = result.popLast() {
                result.append(last + current)
            } else {
                result.append(current)
            }
        }
        return result.isEmpty ? [block] : result
    }

    /// 句末标点后紧跟的引号、括号归上一句；连写的「？！」也算同一句。
    private static func sentences(in block: String) -> [String] {
        var result: [String] = []
        var current = ""
        var ended = false
        for ch in block {
            if ended {
                if enders.contains(ch) || closers.contains(ch) {
                    current.append(ch)
                    continue
                }
                result.append(current)
                current = ""
                ended = false
            }
            current.append(ch)
            if enders.contains(ch) { ended = true }
        }
        if !current.isEmpty { result.append(current) }
        return result
    }
}

File: android/app/src/main/java/com/yizhidao/app/ai/AIAnswerFormatter.kt
package com.yizhidao.app.ai

/**
 * 与 iOS `AIAnswerFormatter` 同规则：AI 常把解读写成一整段，手机上读起来很挤。
 * 只在展示层分段，不动存下来的原文：先按换行切块，再只在句末（。！？…）断开，
 * 累计约 60 字成段，末段过短并回上一段。
 */
object AIAnswerFormatter {
    private const val SOFT_LIMIT = 60
    private const val MIN_TAIL_LENGTH = 24
    private val ENDERS = setOf('。', '！', '？', '…', '!', '?', '.')
    private val CLOSERS = setOf('」', '』', '”', '’', '）', ')', '》', '〉', '】', '、')

    fun paragraphs(raw: String): List<String> = raw
        .replace("\r\n", "\n")
        .replace('\r', '\n')
        .split('\n')
        .map { it.trim() }
        .filter { it.isNotEmpty() }
        .flatMap { paragraphsInBlock(it) }

    private fun paragraphsInBlock(block: String): List<String> {
        if (block.length <= SOFT_LIMIT) return listOf(block)
        val result = mutableListOf<String>()
        val current = StringBuilder()
        for (sentence in sentences(block)) {
            current.append(sentence)
            if (current.length >= SOFT_LIMIT) {
                result.add(current.toString())
                current.setLength(0)
            }
        }
        if (current.isNotEmpty()) {
            if (current.length < MIN_TAIL_LENGTH && result.isNotEmpty()) {
                result[result.lastIndex] = result.last() + current
            } else {
                result.add(current.toString())
            }
        }
        return result.ifEmpty { listOf(block) }
    }

    /** 句末标点后紧跟的引号、括号归上一句；连写的「？！」也算同一句。 */
    private fun sentences(block: String): List<String> {
        val result = mutableListOf<String>()
        val current = StringBuilder()
        var ended = false
        for (ch in block) {
            if (ended) {
                if (ch in ENDERS || ch in CLOSERS) {
                    current.append(ch)
                    continue
                }
                result.add(current.toString())
                current.setLength(0)
                ended = false
            }
            current.append(ch)
            if (ch in ENDERS) ended = true
        }
        if (current.isNotEmpty()) result.add(current.toString())
        return result
    }
}
